### PR TITLE
op-service: use binary search instead of walkback for checkRecentTxs

### PR DESCRIPTION
**Pull request by bitwiseguy**
## Description

To improve the efficiency of the `checkRecentTxs`, this PR changes the search from a simple backwards traversal of blocks one-by-one to a binary search. This is especially helpful when searching large ranges of blocks.

## Tests

Refactored existing `checkRecentTxs` to make it more clear which edge cases we are checking, and their expected outcomes.

## Additional context

The `checkRecentTxs` function is currently used by `op-batcher` and will be used by several superchain-registry validation checks, as mentioned in https://github.com/ethereum-optimism/superchain-registry/issues/400

### CODE REVIEW

File: op-service/eth/transactions_test.go [hunk 1]
```diff
@@ -30,111 +30,105 @@ func (m *MockL1Client) HeaderByNumber(ctx context.Context, number *big.Int) (*ty
...
-		blockConfirms      uint64
-		previousNonceBlock uint64
-		expectedBlockNum   uint64
-		expectedFound      bool
+		name             string
+		currentBlock     uint64
+		blockConfirms    uint64
+		expectedBlockNum uint64
+		expectedFound    bool
+		nonces           map[uint64]uint64 // maps blockNum --> nonce
```
Comment (sebastianst) [inline op-service/eth/transactions_test.go]: If you make this a `[]uint64` it would be a bit more condensed. You can treat the index e.g. as the negative distance from `currentBlock`.

I see that using a map might help us to do exact call assertions after all, see other comment.

File: op-service/eth/transactions.go [hunk 2]
```diff
@@ -69,25 +71,47 @@ func CheckRecentTxs(
...
-	oldestBlock := new(big.Int)
-	oldestBlock.Sub(currentBlock, big.NewInt(int64(depth)))
+	oldestBlock := new(big.Int).Sub(currentBlock, big.NewInt(int64(depth)))
 	previousNonce, err := l1.NonceAt(ctx, addr, oldestBlock)
 	if err != nil {
 		return 0, false, fmt.Errorf("failed to retrieve previous nonce: %w", err)
 	}
 
-	if currentNonce == previousNonce {
+	if currentNonce <= previousNonce {
```
Comment (sebastianst) [inline op-service/eth/transactions.go]: When is this new case `current < prev` hit? Caused by a deep reorg between querying the current and previous nonce? In that case we might want to actually update the `currentNonce` and check again, possibly by just recursively calling `CheckRecentTxs`.

File: op-service/eth/transactions.go [hunk 3]
```diff
@@ -69,25 +71,47 @@ func CheckRecentTxs(
...
+		mid := (low + high) / 2
+		midNonce, err := l1.NonceAt(ctx, addr, new(big.Int).SetUint64(mid))
 		if err != nil {
-			return 0, false, fmt.Errorf("failed to retrieve nonce: %w", err)
+			return 0, false, fmt.Errorf("failed to retrieve nonce at block %d: %w", mid, err)
+		}
+
+		if midNonce > currentNonce {
+			// Catch a reorg that causes inconsistent nonce
+			return oldestBlock.Uint64(), false, nil
```
Comment (sebastianst) [inline op-service/eth/transactions.go]: I'm not sure this is how we want to handle a reorg case. If the mid nonce is _higher_ than the pre-reorg current nonce, it means that there suddenly may be a more recent transaction between oldestBlock and the currentBlock. I think at this point we cannot narrow it down to a smaller range. So maybe what we want to do instead here is to restart a new search for the same range? Easiest would probably be again to just recursively call `CheckRecentTxs` because we'd also need to update the `currentNonce`.
```suggestion
		if midNonce > currentNonce {
			// Restart search on reorg
			return CheckRecentTxs(ctx, l1, depth, addr)
```

File: op-service/eth/transactions.go [hunk 4]
```diff
@@ -69,25 +71,47 @@ func CheckRecentTxs(
...
+				return 0, false, fmt.Errorf("failed to retrieve nonce at block %d: %w", mid, err)
+			}
+
+			if nextBlockNonce == currentNonce {
+				return nextBlockNum, true, nil
+			}
+			low = mid + 1
+		} else {
+			low = mid + 1
 		}
```
Comment (sebastianst) [inline op-service/eth/transactions.go]:
```suggestion
		} else { // if midNonce < currentNonce
			// Check the next block to see if we've found the spot where the nonce increases
			nextBlockNum := mid + 1
			nextBlockNonce, err := l1.NonceAt(ctx, addr, new(big.Int).SetUint64(nextBlockNum))
			if err != nil {
				return 0, false, fmt.Errorf("failed to retrieve nonce at block %d: %w", mid, err)
			}

			if nextBlockNonce == currentNonce {
				return nextBlockNum, true, nil
			}
			low = mid + 1
		}
```
The same block can include multiple transactions from the same sender, so multiple nonce changes are possible. So maybe we want to do this simpler version instead?

File: op-service/eth/transactions_test.go [hunk 5]
```diff
@@ -30,111 +30,105 @@ func (m *MockL1Client) HeaderByNumber(ctx context.Context, number *big.Int) (*ty
...
+			for blockNum, nonce := range tt.nonces {
+				l1Client.On("NonceAt", ctx, common.Address{}, new(big.Int).SetUint64(blockNum)).Return(nonce, nil)
 			}
 
 			blockNum, found, err := CheckRecentTxs(ctx, l1Client, 5, common.Address{})
 			require.NoError(t, err)
-			require.Equal(t, tt.expectedBlockNum, blockNum)
 			require.Equal(t, tt.expectedFound, found)
-
-			l1Client.AssertExpectations(t)
```
Comment (sebastianst) [inline op-service/eth/transactions_test.go:111]: I see that you have to remove the `AssertExpectations` case because not all `NonceAt` calls are performed by the binary search. Maybe we actually do want to be explicit about the expected nonce calls from the binary search? Then it may make sense to use the `nonces: map[uint64]uint64` map instead of a slice after all. Would that work?

File: op-service/eth/transactions_test.go [hunk 6]
```diff
@@ -30,111 +30,105 @@ func (m *MockL1Client) HeaderByNumber(ctx context.Context, number *big.Int) (*ty
...
-					} else if block <= tt.previousNonceBlock {
-						t.Log("previousNonce set at block: ", block)
-						l1Client.On("NonceAt", ctx, common.Address{}, blockBig).Return(previousNonce, nil)
-					} else {
-						t.Log("currentNonce set at block: ", block)
-						l1Client.On("NonceAt", ctx, common.Address{}, blockBig).Return(currentNonce, nil)
-					}
-				}
+			for blockNum, nonce := range tt.nonces {
+				l1Client.On("NonceAt", ctx, common.Address{}, new(big.Int).SetUint64(blockNum)).Return(nonce, nil)
```
Comment (sebastianst) [inline op-service/eth/transactions_test.go]: if you make the map key an `int64` you could use this slightly shorter form
```suggestion
				l1Client.On("NonceAt", ctx, common.Address{}, big.NewInt(blockNum).Return(nonce, nil)
```

File: op-service/eth/transactions_test.go [hunk 7]
```diff
@@ -30,111 +30,105 @@ func (m *MockL1Client) HeaderByNumber(ctx context.Context, number *big.Int) (*ty
...
+			expectedBlockNum: 495,
+			expectedFound:    false,
+			nonces: map[uint64]uint64{
+				495: 5,
+				496: 7,
+				497: 6,
+				498: 6,
+				499: 6,
+				500: 6,
+			},
```
Comment (sebastianst) [inline op-service/eth/transactions_test.go]: If we implement reorg behavior as a restart of the search, this test case becomes a bit more complex to set up, as we'd need to return different nonces on successive calls.

**Review by sebastianst — CHANGES_REQUESTED** [review]

File: op-service/eth/transactions.go (hunk 4 above)
Comment (bitwiseguy) [reply op-service/eth/transactions.go]: Yea good call. Will update

File: op-service/eth/transactions_test.go (hunk 7 above)
Comment (bitwiseguy) [reply op-service/eth/transactions_test.go]: https://github.com/ethereum-optimism/optimism/pull/11232/commits/9f6009764310d8a35f91b2cbe9fa5281f8616eb2

File: op-service/eth/transactions_test.go (hunk 5 above)
Comment (bitwiseguy) [reply op-service/eth/transactions_test.go:111]: https://github.com/ethereum-optimism/optimism/pull/11232/commits/9f6009764310d8a35f91b2cbe9fa5281f8616eb2

File: op-service/eth/transactions.go (hunk 3 above)
Comment (bitwiseguy) [reply op-service/eth/transactions.go]: https://github.com/ethereum-optimism/optimism/pull/11232/commits/9f6009764310d8a35f91b2cbe9fa5281f8616eb2

File: op-service/eth/transactions.go (hunk 2 above)
Comment (bitwiseguy) [reply op-service/eth/transactions.go]: I reverted this change. Instead we should be able to lean on the binary search that follows to catch and handle a reorg. https://github.com/ethereum-optimism/optimism/pull/11232/commits/e13d7497924f72d0fc58b54e2a0a312e7fb14151

File: op-service/eth/transactions_test.go (hunk 1 above)
Comment (bitwiseguy) [reply op-service/eth/transactions_test.go]: Refactored in https://github.com/ethereum-optimism/optimism/pull/11232/commits/9f6009764310d8a35f91b2cbe9fa5281f8616eb2

File: op-service/eth/transactions.go (hunk 2 above)
Comment (sebastianst) [reply op-service/eth/transactions.go]: that makes sense!

**Review by sebastianst — APPROVED** [review]
LGTM ✨
